コメントを付けられるようにする

ryufuta opened the pull request:
レビューをお願いします🙏

コメントの保存に失敗したときは`render`を使いたかったのですが、できなかったので`redirect_to`を使っています。

https://github.com/ryufuta/fjord-books_app-2023/blob/5f2aa4751d6f22db9d14177c50e2c74341fed849/app/controllers/comments_controller.rb#L8-L13

ryufuta committed:
Add Report scaffold associated with User

ryufuta committed:
Set logged in user as report creator

ryufuta committed:
Display username instead of user ID to improve UI

ryufuta committed:
Display report creation date

ryufuta committed:
Allow only author to update and delete reports

ryufuta committed:
Add link to reports in menu

ryufuta committed:
Add Comment model associated with User, Book and Report

ryufuta committed:
Internationalize and localize reports

ryufuta committed:
Improve report UI

ryufuta committed:
Implement displaying comments on report details page

ryufuta committed:
Implement displaying comments on book details page

ryufuta committed:
Eliminate code duplication in displaying comments

ryufuta committed:
Add routing for posting and deleting comments

ryufuta committed:
Implement posting comments on reports

ryufuta committed:
Implement posting comments on books

ryufuta committed:
Implement deleting comments on books and reports

ryufuta committed:
Allow only author to delete comments

ryufuta committed:
Prevent empty comment submission

ryufuta committed:
Localize Comment model

ryufuta committed:
Change time display for reports and comments to improve UI

ryufuta committed:
Remove unused JSON response handling

ryufuta committed:
Remove unused files

betachelsea commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+class CommentsController < ApplicationController
+  before_action :set_commentable, only: %i[create destroy]
```
[nits] 現状で action は `create`, `destroy` の2つだけのようなので、 only を指定しなくても大丈夫そうだと思いました :eyes:

betachelsea commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,36 @@
+
+  def require_authorship
+    redirect_to @commentable unless current_user == @comment.user
+  end
```
[nits] authorship の確認をして、違っていたらリダイレクト、という方法ですね！
この処理が必要とされるのは「意図的に他者のコメントを削除しようとした」時ですね。
こちら、リダイレクトではなくエラーで対処しても問題なさそうなので、「 ログインユーザの持ってるcommentsの中から探して(find)、無ければエラー扱い」という方針でいくと処理が短くなりそうです 💡 

```ruby
# current_userが対象のcommentを持っていなければ ActiveRecord::RecordNotFound の例外エラー
current_user.comments.find(params[:id])
```

betachelsea commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,36 @@
+
+class CommentsController < ApplicationController
+  before_action :set_commentable, only: %i[create destroy]
+  before_action :set_comment, only: %i[destroy]
```
[nits] `only: %i[destroy]` ということで、1箇所に適用するだけなら、メソッドに切り出さずにアクションメソッド内でまとめてしまっても良さそうです :eyes: (うーん、好みの問題もややありそうですが…！)

ryufuta replied:
歓迎要件であるコメントの編集を後で実装するつもりでこのように書いていましたが、結局やらなかった（できなかった）のでご指摘の通り修正しておく方が良かったですね。

betachelsea commented on app/models/user.rb:
```diff
@@ -7,4 +7,7 @@ class User < ApplicationRecord
   end
+
+  has_many :reports, dependent: :destroy
+  has_many :comments, dependent: nil
```
`dependent: nil` は「何もしない」のですが※、DB上では外部キー制約が設定されているため、いざcommentsデータを持つUserを削除しようとした場合、この設定ではエラーが発生しそうです。( Railsは消そうとするけど、DBが消すのを許さないため ) 

```ruby
# 外部キー制約が設定されている様子( db/schema.rb )
add_foreign_key "comments", "users"
```

ぜひ `$ bin/rails console` にて、すでにコメントデータを持っている 任意の Userデータに対して `user.destroy` を実行するとどうなるか、試してみてください 💡 

[MUST] コメントデータを持っているユーザーのデータに対して、`user.destory` を実行してもエラーにならないように修正してもらいたいです！ 🙏 


※参考
[ActiveRecord::Associations::ClassMethods](https://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many)

> :dependent
> ...
> nil do nothing (default).

betachelsea commented on app/views/comments/_form.html.erb:
```diff
@@ -0,0 +1,22 @@
+<%= form_with(model: [commentable, comment]) do |form| %>
+  <% if comment.errors.any? %>
+    <div style="color: red">
```
[nits] 同じ形式のバリデーションエラー表示を何箇所かで見た気がします :eyes: 共通化チャンスかな、と思いました！

betachelsea commented on app/views/reports/_report.html.erb:
```diff
@@ -0,0 +1,22 @@
+  </p>
+
+  <p>
+    <strong><%= User.model_name.human %>:</strong>
```
[nits] Reportの情報の表示ページなので、Report.human_attribute_name で `user` を定義したものを呼び出したほうが、あとから取り回しが効きやすいと思いました 💡  例えば、同じユーザーを示すのでも、モデルによっては「ユーザー」「著者」「投稿者」などと呼び方が変わる、というのはありそうです。そういう時の場合、呼び出し元(今回はReport)のほうから User を呼ぶ時は なんと呼称するか？ を別で定義しておくと、他の場所に影響を出さずに変更を入れられて便利です！

betachelsea commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,68 @@
+
+  # GET /reports or /reports.json
+  def index
+    @reports = Report.all
```
[MUST] ページネーションが欲しいです！ 何かのデータを `all` したくなった時は、一度立ち止まる癖をつけておくといいです。テスト環境とかだと、せいぜい 50〜100個ほどしかデータが入っていないので問題なく見えますが、実際に稼働しているWebサービスだとウン千・ウン万・ウン百万のデータを扱ってるのがあたりまえ…、みたいな世界観になります。〇〇を一覧にしよう、と思ったときは、ページネーションの機能も必要だなぁ、と関連付けて思い出せるようにしていきましょう！ 💡

betachelsea commented on config/routes.rb:
```diff
@@ -1,7 +1,14 @@
 Rails.application.routes.draw do
+  resources :reports
```
[MUST] 8行目〜で書かれている `resources :reports do (〜end)` で、同じルーティング設定ができていますので、別途単独の `resources :reports` は書かなくても大丈夫です！ 重複になってしまっているので、こちらの行は消してほしいです。booksについても同じ状況みたいですね :eyes:

betachelsea commented on db/migrate/20240705091507_create_reports.rb:
```diff
@@ -0,0 +1,11 @@
+  def change
+    create_table :reports do |t|
+      t.string :title
+      t.text :content
```
[MUST] 日報ですが、タイトル・本文が空でもどんどん投稿できる様子でした :eyes: こちら、タイトル・本文は中身が無いと投稿できないようにしてみてもらいたいです :pray: また、コメントの入力フォームはHTML側で必須チェックをされているようですが、日報のタイトル・本文の必須チェックはバリデーションを活用してみてもらいたいです！
また、DB側(migrationの設定時)も `null: false` の指定をしてみてほしいです！

betachelsea commented on test/system/reports_test.rb:
```diff
@@ -0,0 +1,47 @@
+
+class ReportsTest < ApplicationSystemTestCase
+  setup do
+    @report = reports(:one)
```
[nits] reportsのfixturesが見当たらないようです。こちらは自動的に作成されたテストファイルでしょうか？ テストは今回のプラクティスの範囲外なので、無くても大丈夫ですが、利用予定のないテストファイルはPullRequest内からは除外をお願いします 💡

ryufuta replied:
不要なファイルはすべて削除したつもりでしたが、漏れていました🙇🏻‍♂️

betachelsea reviewed (commented):
提出お疲れさまです！ 🍵  コメントしました〜。

ryufuta committed:
Remove unused files

ryufuta committed:
Remove useless code

ryufuta committed:
Remove redundant code

ryufuta committed:
Combine set_comment and require_authorship methods

- For brevity
- Better to return error rather than redirect for unauthorized delete attempts

ryufuta committed:
Combine destroy and set_comment methods for brevity

ryufuta committed:
Change report permission check logic

To match comments_controller.rb changes

ryufuta committed:
Move set_report into show method for brevity

ryufuta committed:
Add translations of user attributes for each model

To improve maintainability

ryufuta committed:
Fix bug that prevents deleting users with comments

ryufuta committed:
Fix bug preventing display of deleted users' comments

ryufuta committed:
Remove code duplication in displaying form errors

ryufuta committed:
Add pagination to reports page

ryufuta committed:
Add pagination to comments

ryufuta committed:
Add validation to report's title and content

ryufuta committed:
Add NOT NULL constraints to title and content in reports table

betachelsea commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,63 @@
+  # GET /reports/1/edit
+  def edit; end
+
+  # POST /reports or /reports.json
```
[nits] おそらく最初にgenerateした時に json があったかもしれませんが、今の実装にはないので、 `or /reports.json` のコメントは不要そうですね〜 👀

betachelsea commented on app/views/books/_form.html.erb:
```diff
@@ -1,15 +1,5 @@
-      </ul>
-    </div>
-  <% end %>
+  <%= render "application/form_error", { model_instance: book, model_class: Book } %>
```
[TIPS] model_class は、 `model_instance.class` とすると取得できるので、引数を1つ減らせそうです！ 💡

betachelsea commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,63 @@
+
+  # GET /reports or /reports.json
+  def index
+    @reports = Report.order(:id).page(params[:page])
```
[nits] 万が一ですが、 order の仕様がRailsのActiveRecord側で変わった場合に意図しない順番になる可能性があるので、降順/昇順も明示的に表記するのをオススメします！ > `order(:id)`  
他にも comments の order を決めてる箇所とかもあるみたいですね。全般的にorderするときは降順/昇順を明示しておくと、後から読んだ時にもわかりやすくて良いと思います 💡

betachelsea commented on db/migrate/20240708091558_create_comments.rb:
```diff
@@ -0,0 +1,11 @@
+  def change
+    create_table :comments do |t|
+      t.text :content
+      t.references :user, foreign_key: { on_delete: :nullify }
```
ユーザーをnullにしてコメントを残す、ということが明示的になりましたね 👍  消した時の表示もちゃんと「削除されたユーザ」と出るようになっていてGoodですね！ その辺りのケアができていないと、ページを見ようとした時に存在しない情報を参照しにいこうとして例外エラーになってしまう、とかはあるあるなので、要注意ポイントだったと思います。LGTM！

betachelsea commented on app/views/comments/_comment.html.erb:
```diff
@@ -0,0 +1,21 @@
+    <% if comment.user.nil? %>
+      <%= t('views.common.deleted_record', name: Comment.human_attribute_name(:user).downcase) %>
+    <% else %>
+      <%= current_user_name(comment.user) %>
```
[TIPS] 「userを失ったコメント」というのは、ここ以外でも扱うところがあると思います。今回はこのviewの修正のみで収まりそうなのでよかったですが、もし色んなviewで使われていたりする場合は、モデル側で判断/出力内容を調整したり、この `current_user_name` メソッドとかで調整したり、とかの手段もあります。豆知識まで 💡

betachelsea reviewed (approved):
諸々修正ありがとうございました！ LGTMです 👍 
いくつか補足コメントさせていただきました〜 ✍️
